File: internal/forms/federal/f1040.go
package federal

import (
	"math"

	"taxpilot/internal/forms"
	"taxpilot/pkg/taxmath"
)

func init() { forms.RegisterForm(F1040) }

// F1040 returns the FormDef for Form 1040 — U.S. Individual Income Tax Return.
// This is a simplified MVP covering a single (or multiple) W-2 filer with
// standard deduction. Additional income sources and itemized deductions will
// be added in future iterations.
func F1040() *forms.FormDef {
	return &forms.FormDef{
		ID:           forms.FormF1040,
		Name:         "U.S. Individual Income Tax Return",
		Jurisdiction: forms.Federal,
		TaxYears:      []int{2025},
		QuestionGroup: forms.GroupPersonal,
		QuestionOrder: 1,
		Fields: []forms.FieldDef{
			// --- Identification ---
			{
				Line:    forms.LineFilingStatus,
				Type:    forms.UserInput,
				ValueType: forms.StringValue,
				Label:   "Filing status",
				Prompt:  "What is your filing status?",
				Options: forms.FilingStatusOptions,
			},
			{
				Line:   forms.LineFirstName,
				Type:   forms.UserInput,
				ValueType: forms.StringValue,
				Label:  "First name",
				Prompt: "What is your first name?",
			},
			{
				Line:   forms.LineLastName,
				Type:   forms.UserInput,
				ValueType: forms.StringValue,
				Label:  "Last name",
				Prompt: "What is your last name?",
			},
			{
				Line:   forms.LineSSN,
				Type:   forms.UserInput,
				ValueType: forms.StringValue,
				Label:  "Social Security number",
				Prompt: "What is your Social Security number (XXX-XX-XXXX)?",
			},

			// --- Income ---

			// Foreign wages not reported on a US W-2 (e.g., from a foreign employer)
			{
				Line:      forms.LineForeignWages,
				Type:      forms.UserInput,
				Label:     "Foreign wages (not on W-2)",
				Prompt:    "Enter wages from foreign employers not reported on a US W-2, converted to USD:",
			},
			// Foreign employer description
			{
				Line:      forms.LineForeignEmployer,
				Type:      forms.UserInput,
				ValueType: forms.StringValue,
				Label:     "Foreign employer(s)",
				Prompt:    "Describe the foreign employer(s) (e.g., \"Volvo AB, Sweden\"):",
			},

			// Line 1a: Wages from W-2s + foreign wages
			{
				Line:      "1a",
				Type:      forms.Computed,
				Label:     "Wages, salaries, tips (W-2 + foreign)",
				DependsOn: []string{forms.W2WildcardWages, forms.F1040ForeignWages},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAll(forms.W2WildcardWages) + dv.Get(forms.F1040ForeignWages)
				},
			},
			// Line 1z: Total from adding lines 1a through 1h (MVP: only 1a)
			{
				Line:      "1z",
				Type:      forms.Computed,
				Label:     "Total from W-2s and other wage sources",
				DependsOn: []string{forms.F1040Line1a},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F1040Line1a)
				},
			},
			// Line 2a: Tax-exempt interest (from 1099-INT Box 8)
			{
				Line:      "2a",
				Type:      forms.Computed,
				Label:     "Tax-exempt interest",
				DependsOn: []string{forms.F1099INTWildcardTaxExempt},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAll(forms.F1099INTWildcardTaxExempt)
				},
			},
			// Line 2b: Taxable interest (from Schedule B line 4)
			{
				Line:      "2b",
				Type:      forms.Computed,
				Label:     "Taxable interest",
				DependsOn: []string{forms.SchedBLine4},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.SchedBLine4)
				},
			},
			// Line 3a: Qualified dividends (from 1099-DIV Box 1b)
			{
				Line:      "3a",
				Type:      forms.Computed,
				Label:     "Qualified dividends",
				DependsOn: []string{forms.F1099DIVWildcardQualified},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAll(forms.F1099DIVWildcardQualified)
				},
			},
			// Line 3b: Ordinary dividends (from Schedule B line 6)
			{
				Line:      "3b",
				Type:      forms.Computed,
				Label:     "Ordinary dividends",
				DependsOn: []string{forms.SchedBLine6},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.SchedBLine6)
				},
			},
			// Line 7: Capital gain or (loss) from Schedule 1 line 7
			{
				Line:      "7",
				Type:      forms.Computed,
				Label:     "Capital gain or (loss)",
				DependsOn: []string{forms.Sched1Line7},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.Sched1Line7)
				},
			},
			// Line 8: Other income from Schedule 1 line 10
			{
				Line:      "8",
				Type:      forms.Computed,
				Label:     "Other income from Schedule 1",
				DependsOn: []string{forms.Sched1Line10},
				Compute: func(dv forms.DepValues) float64 {
					// Schedule 1 line 10 minus capital gains (already on line 7)
					return dv.Get(forms.Sched1Line10) - dv.Get(forms.Sched1Line7)
				},
			},
			// Line 9: Total income
			{
				Line:      "9",
				Type:      forms.Computed,
				Label:     "Total income",
				DependsOn: []string{forms.F1040Line1z, forms.F1040Line2b, forms.F1040Line3b, forms.F1040Line7, forms.F1040Line8},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F1040Line1z) +
						dv.Get(forms.F1040Line2b) +
						dv.Get(forms.F1040Line3b) +
						dv.Get(forms.F1040Line7) +
						dv.Get(forms.F1040Line8)
				},
			},
			// Line 10: Adjustments to income (from Schedule 1 line 26)
			{
				Line:      "10",
				Type:      forms.Computed,
				Label:     "Adjustments to income",
				DependsOn: []string{forms.Sched1Line26},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.Sched1Line26)
				},
			},
			// Line 11: Adjusted gross income (AGI)
			{
				Line:      "11",
				Type:      forms.Computed,
				Label:     "Adjusted gross income",
				DependsOn: []string{forms.F1040Line9, forms.F1040Line10},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F1040Line9) - dv.Get(forms.F1040Line10)
				},
			},

			// --- Deductions ---

			// Line 12: Deduction — larger of standard deduction or itemized (Schedule A)
			{
				Line:      "12",
				Type:      forms.Computed,
				Label:     "Standard deduction or itemized deductions",
				DependsOn: []string{forms.F1040FilingStatus, forms.SchedALine17},
				Compute: func(dv forms.DepValues) float64 {
					fs := taxmath.FilingStatus(dv.GetString(forms.F1040FilingStatus))
					standard := taxmath.GetStandardDeduction(dv.TaxYear(), taxmath.Federal, fs)
					itemized := dv.Get(forms.SchedALine17)
					return math.Max(standard, itemized)
				},
			},
			// Line 13: Qualified business income deduction (from Form 8995)
			{
				Line:      "13",
				Type:      forms.Computed,
				Label:     "Qualified business income deduction",
				DependsOn: []string{forms.F8995Line10},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F8995Line10)
				},
			},
			// Line 14: Total deductions
			{
				Line:      "14",
				Type:      forms.Computed,
				Label:     "Total deductions",
				DependsOn: []string{forms.F1040Line12, forms.F1040Line13},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F1040Line12) + dv.Get(forms.F1040Line13)
				},
			},
			// Line 15: Taxable income
			{
				Line:      "15",
				Type:      forms.Computed,
				Label:     "Taxable income",
				DependsOn: []string{forms.F1040Line11, forms.F1040Line14},
				Compute: func(dv forms.DepValues) float64 {
					return math.Max(0, dv.Get(forms.F1040Line11)-dv.Get(forms.F1040Line14))
				},
			},

			// --- Tax computation ---

			// Line 16: Tax
			// When FEIE (Form 2555) is claimed, uses the "stacking" method:
			// tax is computed at the rate that would apply if the excluded
			// income were still included, preventing bracket manipulation.
			{
				Line:      "16",
				Type:      forms.Computed,
				Label:     "Tax",
				DependsOn: []string{forms.F1040Line15, forms.F1040FilingStatus, forms.F2555TotalExclusion},
				Compute: func(dv forms.DepValues) float64 {
					fs := taxmath.FilingStatus(dv.GetString(forms.F1040FilingStatus))
					taxableIncome := dv.Get(forms.F1040Line15)
					excludedIncome := dv.Get(forms.F2555TotalExclusion)
					if excludedIncome > 0 {
						return taxmath.ComputeTaxWithStacking(taxableIncome, excludedIncome, fs, dv.TaxYear(), taxmath.Federal)
					}
					return taxmath.ComputeTax(taxableIncome, fs, dv.TaxYear(), taxmath.Federal)
				},
			},
			// Line 17: Amount from Schedule 2 Part I (AMT, etc.)
			{
				Line:      "17",
				Type:      forms.Computed,
				Label:     "Amount from Schedule 2, Part I",
				DependsOn: []string{forms.Sched2Line3},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.Sched2Line3)
				},
			},
			// Line 20: Amount from Schedule 3 Part I (nonrefundable credits)
			{
				Line:      "20",
				Type:      forms.Computed,
				Label:     "Amount from Schedule 3, Part I",
				DependsOn: []string{forms.Sched3Line8},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.Sched3Line8)
				},
			},
			// Line 22: Tax after credits (line 16 + 17 - 20, but not less than 0)
			{
				Line:      "22",
				Type:      forms.Computed,
				Label:     "Tax after credits",
				DependsOn: []string{forms.F1040Line16, forms.F1040Line17, forms.F1040Line20},
				Compute: func(dv forms.DepValues) float64 {
					return math.Max(0, dv.Get(forms.F1040Line16)+dv.Get(forms.F1040Line17)-dv.Get(forms.F1040Line20))
				},
			},
			// Line 23: Other taxes from Schedule 2 Part II
			{
				Line:      "23",
				Type:      forms.Computed,
				Label:     "Other taxes from Schedule 2",
				DependsOn: []string{forms.Sched2Line21},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.Sched2Line21)
				},
			},
			// Line 24: Total tax (line 22 + line 23)
			{
				Line:      "24",
				Type:      forms.Computed,
				Label:     "Total tax",
				DependsOn: []string{forms.F1040Line22, forms.F1040Line23},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F1040Line22) + dv.Get(forms.F1040Line23)
				},
			},

			// --- Payments ---

			// Line 25a: Federal income tax withheld from W-2s
			{
				Line:      "25a",
				Type:      forms.Computed,
				Label:     "Federal income tax withheld from W-2s",
				DependsOn: []string{forms.W2WildcardFedTaxWH},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAll(forms.W2WildcardFedTaxWH)
				},
			},
			// Line 25b: Federal income tax withheld from 1099s
			{
				Line:      "25b",
				Type:      forms.Computed,
				Label:     "Federal income tax withheld from 1099s",
				DependsOn: []string{forms.F1099INTWildcardFedTaxWH, forms.F1099DIVWildcardFedTaxWH, forms.F1099NECWildcardFedTaxWH, forms.F1099BWildcardFedTaxWH},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAll(forms.F1099INTWildcardFedTaxWH) +
						dv.SumAll(forms.F1099DIVWildcardFedTaxWH) +
						dv.SumAll(forms.F1099NECWildcardFedTaxWH) +
						dv.SumAll(forms.F1099BWildcardFedTaxWH)
				},
			},
			// Line 25d: Total federal tax withheld
			{
				Line:      "25d",
				Type:      forms.Computed,
				Label:     "Total federal tax withheld",
				DependsOn: []string{forms.F1040Line25a, forms.F1040Line25b},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F1040Line25a) + dv.Get(forms.F1040Line25b)
				},
			},
			// Line 31: Other payments from Schedule 3 (estimated tax, etc.)
			{
				Line:      "31",
				Type:      forms.Computed,
				Label:     "Other payments from Schedule 3",
				DependsOn: []string{forms.Sched3Line15},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.Sched3Line15)
				},
			},
			// Line 33: Total payments
			{
				Line:      "33",
				Type:      forms.Computed,
				Label:     "Total payments",
				DependsOn: []string{forms.F1040Line25d, forms.F1040Line31},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F1040Line25d) + dv.Get(forms.F1040Line31)
				},
			},

			// --- Refund or Amount Owed ---

			// Line 34: Overpayment (refund)
			{
				Line:      "34",
				Type:      forms.Computed,
				Label:     "Overpayment (refund)",
				DependsOn: []string{forms.F1040Line33, forms.F1040Line24},
				Compute: func(dv forms.DepValues) float64 {
					return math.Max(0, dv.Get(forms.F1040Line33)-dv.Get(forms.F1040Line24))
				},
			},
			// Line 37: Amount you owe
			{
				Line:      "37",
				Type:      forms.Computed,
				Label:     "Amount you owe",
				DependsOn: []string{forms.F1040Line24, forms.F1040Line33},
				Compute: func(dv forms.DepValues) float64 {
					return math.Max(0, dv.Get(forms.F1040Line24)-dv.Get(forms.F1040Line33))
				},
			},
		},
	}
}

File: api/src/forms/federal/f1040.rs
use crate::domain::builders::*;
use crate::domain::field::*;
use crate::domain::form::*;
use crate::domain::taxmath;

pub fn form_1040() -> FormDef {
    FormDef {
        id: FORM_F1040.to_string(),
        name: "U.S. Individual Income Tax Return".to_string(),
        jurisdiction: Jurisdiction::Federal,
        tax_years: vec![2024, 2025, 2026],
        question_group: "personal".to_string(),
        question_order: 1,
        fields: vec![
            // --- Identification ---
            enum_field(
                "filing_status",
                "Filing status",
                "What is your filing status?",
                vec!["single", "mfj", "mfs", "hoh", "qss"],
            ),
            string_input_field("first_name", "First name", "What is your first name?"),
            string_input_field("last_name", "Last name", "What is your last name?"),
            string_input_field(
                "ssn",
                "Social Security number",
                "What is your Social Security number (XXX-XX-XXXX)?",
            ),
            // --- Income ---
            input_field(
                "foreign_wages",
                "Foreign wages (not on W-2)",
                "Enter wages from foreign employers not reported on a US W-2, converted to USD:",
            ),
            string_input_field(
                "foreign_employer",
                "Foreign employer(s)",
                "Describe the foreign employer(s) (e.g., \"Volvo AB, Sweden\"):",
            ),
            // Line 1a: Wages from W-2s + foreign wages
            {
                let deps = vec![
                    W2_WILDCARD_WAGES.to_string(),
                    F1040_FOREIGN_WAGES.to_string(),
                ];
                FieldDef::new_computed("1a", "Wages, salaries, tips (W-2 + foreign)", deps, Box::new(|dv: &DepValues| {
                    dv.sum_all(W2_WILDCARD_WAGES) + dv.get(F1040_FOREIGN_WAGES)
                }))
            },
            // Line 1z
            ref_field("1z", "Total from W-2s and other wage sources", F1040_LINE_1A),
            // Line 2a: Tax-exempt interest
            wildcard_sum_field("2a", "Tax-exempt interest", F1099_INT_WILDCARD_TAX_EXEMPT),
            // Line 2b: Taxable interest
            ref_field("2b", "Taxable interest", SCHED_B_LINE_4),
            // Line 3a: Qualified dividends
            wildcard_sum_field("3a", "Qualified dividends", F1099_DIV_WILDCARD_QUALIFIED),
            // Line 3b: Ordinary dividends
            ref_field("3b", "Ordinary dividends", SCHED_B_LINE_6),
            // Line 7: Capital gain or (loss)
            ref_field("7", "Capital gain or (loss)", SCHED_1_LINE_7),
            // Line 8: Other income from Schedule 1
            {
                let deps = vec![SCHED_1_LINE_10.to_string()];
                FieldDef::new_computed("8", "Other income from Schedule 1", deps, Box::new(|dv: &DepValues| {
                    dv.get(SCHED_1_LINE_10) - dv.get(SCHED_1_LINE_7)
                }))
            },
            // Line 9: Total income
            sum_field("9", "Total income", vec![
                F1040_LINE_1Z, F1040_LINE_2B, F1040_LINE_3B, F1040_LINE_7, F1040_LINE_8,
            ]),
            // Line 10: Adjustments to income
            ref_field("10", "Adjustments to income", SCHED_1_LINE_26),
            // Line 11: Adjusted gross income (AGI)
            max_zero_field("11", "Adjusted gross income", F1040_LINE_9, F1040_LINE_10),
            // Line 12: Deduction - larger of standard or itemized
            {
                let deps = vec![
                    F1040_FILING_STATUS.to_string(),
                    SCHED_A_LINE_17.to_string(),
                ];
                FieldDef::new_computed("12", "Standard deduction or itemized deductions", deps, Box::new(|dv: &DepValues| {
                    let fs_str = dv.get_string(F1040_FILING_STATUS);
                    let fs = taxmath::FilingStatus::from_str_code(&fs_str).unwrap_or(taxmath::FilingStatus::Single);
                    let standard = taxmath::get_standard_deduction(dv.tax_year(), taxmath::JurisdictionType::Federal, fs);
                    let itemized = dv.get(SCHED_A_LINE_17);
                    standard.max(itemized)
                }))
            },
            // Line 13: QBI deduction
            ref_field("13", "Qualified business income deduction", F8995_LINE_10),
            // Line 14: Total deductions
            sum_field("14", "Total deductions", vec![F1040_LINE_12, F1040_LINE_13]),
            // Line 15: Taxable income
            max_zero_field("15", "Taxable income", F1040_LINE_11, F1040_LINE_14),
            // Line 16: Tax (with FEIE stacking)
            {
                let deps = vec![
                    F1040_LINE_15.to_string(),
                    F1040_FILING_STATUS.to_string(),
                    F2555_TOTAL_EXCLUSION.to_string(),
                ];
                FieldDef::new_computed("16", "Tax", deps, Box::new(|dv: &DepValues| {
                    let fs_str = dv.get_string(F1040_FILING_STATUS);
                    let fs = taxmath::FilingStatus::from_str_code(&fs_str).unwrap_or(taxmath::FilingStatus::Single);
                    let taxable_income = dv.get(F1040_LINE_15);
                    let excluded_income = dv.get(F2555_TOTAL_EXCLUSION);
                    if excluded_income > 0.0 {
                        taxmath::compute_tax_with_stacking(taxable_income, excluded_income, fs, dv.tax_year(), taxmath::JurisdictionType::Federal)
                    } else {
                        taxmath::compute_tax(taxable_income, fs, dv.tax_year(), taxmath::JurisdictionType::Federal)
                    }
                }))
            },
            // Line 17
            ref_field("17", "Amount from Schedule 2, Part I", SCHED_2_LINE_3),
            // Line 20
            ref_field("20", "Amount from Schedule 3, Part I", SCHED_3_LINE_8),
            // Line 22: Tax after credits
            {
                let deps = vec![
                    F1040_LINE_16.to_string(),
                    F1040_LINE_17.to_string(),
                    F1040_LINE_20.to_string(),
                ];
                FieldDef::new_computed("22", "Tax after credits", deps, Box::new(|dv: &DepValues| {
                    (dv.get(F1040_LINE_16) + dv.get(F1040_LINE_17) - dv.get(F1040_LINE_20)).max(0.0)
                }))
            },
            // Line 23
            ref_field("23", "Other taxes from Schedule 2", SCHED_2_LINE_21),
            // Line 24: Total tax
            sum_field("24", "Total tax", vec![F1040_LINE_22, F1040_LINE_23]),
            // --- Payments ---
            // Line 25a
            wildcard_sum_field("25a", "Federal income tax withheld from W-2s", W2_WILDCARD_FED_TAX_WH),
            // Line 25b
            {
                let deps = vec![
                    F1099_INT_WILDCARD_FED_TAX_WH.to_string(),
                    F1099_DIV_WILDCARD_FED_TAX_WH.to_string(),
                    F1099_NEC_WILDCARD_FED_TAX_WH.to_string(),
                    F1099_B_WILDCARD_FED_TAX_WH.to_string(),
                ];
                FieldDef::new_computed("25b", "Federal income tax withheld from 1099s", deps, Box::new(|dv: &DepValues| {
                    dv.sum_all(F1099_INT_WILDCARD_FED_TAX_WH)
                        + dv.sum_all(F1099_DIV_WILDCARD_FED_TAX_WH)
                        + dv.sum_all(F1099_NEC_WILDCARD_FED_TAX_WH)
                        + dv.sum_all(F1099_B_WILDCARD_FED_TAX_WH)
                }))
            },
            // Line 25d
            sum_field("25d", "Total federal tax withheld", vec![F1040_LINE_25A, F1040_LINE_25B]),
            // Line 31
            ref_field("31", "Other payments from Schedule 3", SCHED_3_LINE_15),
            // Line 33
            sum_field("33", "Total payments", vec![F1040_LINE_25D, F1040_LINE_31]),
            // --- Refund or Amount Owed ---
            // Line 34
            max_zero_field("34", "Overpayment (refund)", F1040_LINE_33, F1040_LINE_24),
            // Line 37
            max_zero_field("37", "Amount you owe", F1040_LINE_24, F1040_LINE_33),
        ],
    }
}
